### PR TITLE
Added documentation about configuring 2fa for users on 9.5+

### PR DESCRIPTION
**Pull request by bergmania**
When 9.5 is released, it is easier to configure 2fa for users..

### CODE REVIEW

**Comment by eshanrnh** [comment]
Thanks for the PR @bergmania 💪

We'll make sure this is reviewed before the 9.5 release.

**Comment by bergmania** [comment]
The sooner the better. 9.5RC goes out tomorrow 🙈

File: Reference/Security/Two-factor-authentication/index.md [hunk 1]
```diff
@@ -184,6 +194,310 @@ If the `setupData` is `null` for the specified `providerName` it means the provi
...
+        /// </summary>
+        /// <remarks>Called to confirm the setup of two factor on the user. In this case we confirm in the same way as we login by validating the PIN.</remarks>
+        public bool ValidateTwoFactorSetup(string secret, string token) => ValidateTwoFactorPIN(secret, token);
+    }
+}
+```
+
+First, we create a model with the information required to set up the 2FA provider and then we implement the `ITwoFactorProvider` with the use of the `TwoFactorAuthenticator` from the GoogleAuthenticator NuGet package.
+
+Now we need to register the `UmbracoUserAppAuthenticator` implementation and the view to show to setup this provider.
```
Comment (eshanrnh) [inline Reference/Security/Two-factor-authentication/index.md:297]: set up

File: Reference/Security/Two-factor-authentication/index.md [hunk 2]
```diff
@@ -184,6 +194,310 @@ If the `setupData` is `null` for the specified `providerName` it means the provi
...
+This can be done on the `IUmbracoBuilder` in your startup or in a composer.
```
Comment (eshanrnh) [inline Reference/Security/Two-factor-authentication/index.md:298]: or a composer.

File: Reference/Security/Two-factor-authentication/index.md [hunk 3]
```diff
@@ -184,6 +194,310 @@ If the `setupData` is `null` for the specified `providerName` it means the provi
...
+
+    </form>
+
+  </umb-editor-view>
+
+</div>
+```
+As this view uses an angular controller we need to create that class and configure it in package.manifest.
+
+In package.manifest we point choose a path to the angular controller.
```
Comment (eshanrnh) [inline Reference/Security/Two-factor-authentication/index.md:395]: In package.manifest, we point to choose

File: Reference/Security/Two-factor-authentication/index.md [hunk 4]
```diff
@@ -184,6 +194,310 @@ If the `setupData` is `null` for the specified `providerName` it means the provi
...
+
+      </umb-editor-footer>
+
+    </form>
+
+  </umb-editor-view>
+
+</div>
+```
+As this view uses an angular controller we need to create that class and configure it in package.manifest.
```
Comment (eshanrnh) [inline Reference/Security/Two-factor-authentication/index.md:393]: As this view uses an angular controller,

File: Reference/Security/Two-factor-authentication/index.md [hunk 5]
```diff
@@ -184,6 +194,310 @@ If the `setupData` is `null` for the specified `providerName` it means the provi
...
+
+      init();
+    }
+  ];
+
+  angular.module("umbraco").controller("CustomCode.TwoFactorProviderGoogleAuthenticator", googleTwoFactorProviderCtrl);
+})();
+```
+
+At this point the 2FA is actually active, but no users have setup 2FA yet.
```
Comment (eshanrnh) [inline Reference/Security/Two-factor-authentication/index.md:475]: At this point, the 2FA is active, but no users have set up 2FA yet.

File: Reference/Security/Two-factor-authentication/index.md [hunk 6]
```diff
@@ -184,6 +194,310 @@ If the `setupData` is `null` for the specified `providerName` it means the provi
...
+
+Each user will now a able to enable the configured 2fa providers on their user. This can be done from the user panel by clicking the user avatar.
```
Comment (eshanrnh) [inline Reference/Security/Two-factor-authentication/index.md:477]: be able

File: Reference/Security/Two-factor-authentication/index.md [hunk 7]
```diff
@@ -184,6 +194,310 @@ If the `setupData` is `null` for the specified `providerName` it means the provi
...
+
+![User panel](images/user-panel.png)
+
+When clicking the `Configure Two-Factor` button, a new panel is shown, listing all enalbed two-factor providers.
```
Comment (eshanrnh) [inline Reference/Security/Two-factor-authentication/index.md:481]: enabled

File: Reference/Security/Two-factor-authentication/index.md [hunk 8]
```diff
@@ -184,6 +194,310 @@ If the `setupData` is `null` for the specified `providerName` it means the provi
...
+
+![Configure 2fa](images/configure-2fa.png)
+
+When clicking `Enable` on one of these, the configured view for the specific provider will be shown
+
+![Enable 2fa](images/enable-2fa.png)
+
+When the authenticator is enabled correct a disable button is shown instead.
```
Comment (eshanrnh) [inline Reference/Security/Two-factor-authentication/index.md:489]: correctly,

File: Reference/Security/Two-factor-authentication/index.md [hunk 9]
```diff
@@ -184,6 +194,310 @@ If the `setupData` is `null` for the specified `providerName` it means the provi
...
+
+![Disable 2fa](images/disable-2fa.png)
+
+To disable the two on your own user, it is required to enter the verification code, otherwise admins are allowed to disable providers on other users.
```
Comment (eshanrnh) [inline Reference/Security/Two-factor-authentication/index.md:493]: otherwise,

File: Reference/Security/Two-factor-authentication/index.md [hunk 10]
```diff
@@ -184,6 +194,310 @@ If the `setupData` is `null` for the specified `providerName` it means the provi
...
+
+![Verify disable](images/verify-disable.png)
+
+While the 2fa is enabled, the user will be presented this screen after entering username and password.
```
Comment (eshanrnh) [inline Reference/Security/Two-factor-authentication/index.md:497]: with this screen after entering the username

File: Reference/Security/Two-factor-authentication/index.md [hunk 11]
```diff
@@ -184,6 +194,310 @@ If the `setupData` is `null` for the specified `providerName` it means the provi
...
+![Verify disable](images/login-2fa.png)
+
+
+## Notification when 2FA is requested for a user
+When a 2FA login is requested for a user, the `UserTwoFactorRequestedNotification` is published.
+This notification can also be used to send the user a one-time password on e-mail or phone, even though these 2FA types are [not considered as secure](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/mfa?view=aspnetcore-6.0#mfa-sms) as App Authentication, it is still a massive improvement compared to no 2FA.
```
Comment (eshanrnh) [inline Reference/Security/Two-factor-authentication/index.md:503]: via e-mail or phone, even though these 2FA types are [not considered secure]

**Review by eshanrnh — COMMENTED** [review]
Looks good 💪 Just couple of comments 🙂

**Review by eshanrnh — COMMENTED** [review]
A couple of suggestions.

**Comment by hemraker** [comment]
@eshanrnh great suggestions!!!! Merging in so it can be used for the RC release notes 😀

**Pull request merged**